cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/main/MessageUtil.java

```diff
@@ -0,0 +1,219 @@
+  private static final Logger LOG = LoggerFactory.getLogger(MessageUtil.class);
+  private static final SampledLogging SAMPLED_LOGGING = new SampledLogging(LOG, 100);
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(Filter.class, new FilterDeserializer())
+    .create();
+  private static final Type MAP_TYPE =
+    new TypeToken<Map<String, String>>() { }.getType();
+  // Number of 100ns intervals since 15 October 1582 00:00:000000000 until UNIX epoch
+  private static final long NUM_100NS_INTERVALS_SINCE_UUID_EPOCH = 0x01b21dd213814000L;
+  // Multiplier to convert millisecond into 100ns
+  private static final long HUNDRED_NANO_MULTIPLIER = 10000;
+
+  private MessageUtil() {
+
+  }
+
+  /**
+   * For each of the namespace at the base location, find the latest file under that namespace and from
+   * the last written record, find the latest messageId for that namespace,
+   * return the overall max messageId across the namespaces, if no namespace is found, return null
+   * @param baseLocation base location for all the namespaces under reporting fileset*
+   * @return messageId
+   * @throws InterruptedException
+   */
+  @Nullable
+  public static String findMessageId(Location baseLocation) throws InterruptedException {
+    List<Location> namespaces = listLocationsWithRetry(baseLocation);
+    byte[] messageId = Bytes.EMPTY_BYTE_ARRAY;
+    String resultMessageId = null;
+    for (Location namespaceLocation : namespaces) {
+      // TODO keep trying with earlier file, if the latest file is empty
+      Optional<Location> latest = findLatestFileLocation(namespaceLocation);
+      if (latest.isPresent()) {
+        String messageIdString = getLatestMessageIdFromFile(latest.get());
+        if (messageIdString != null) {
+          if (Bytes.compareTo(Bytes.fromHexString(messageIdString), messageId) > 0) {
+            messageId = Bytes.fromHexString(messageIdString);
+            resultMessageId = messageIdString;
+          }
+        }
+      }
+    }
+    return resultMessageId;
+  }
+
+  /**
+   * Based on the {@link Message} and its ProgramStatus,
+   * construct by setting the fields of {@link ProgramRunInfo} and return that.
+   * @param message TMS message
+   * @return {@link ProgramRunInfo}
+   */
+  public static ProgramRunInfo constructAndGetProgramRunIdFields(Message message) {
+    Notification notification = GSON.fromJson(message.getPayloadAsString(), Notification.class);
+    ProgramRunInfo programRunInfo =
+      GSON.fromJson(notification.getProperties().get(Constants.Notification.PROGRAM_RUN_ID), ProgramRunInfo.class);
+    programRunInfo.setMessageId(message.getId());
+
+    String programStatus = notification.getProperties().get(Constants.Notification.PROGRAM_STATUS);
+    programRunInfo.setStatus(programStatus);
+
+    switch (programStatus) {
+      case Constants.Notification.Status.STARTING:
+        long startTime = getTime(programRunInfo.getRun(), TimeUnit.MILLISECONDS);
```

Comment:
not sure if we should still use MILLISECONDS since the time unit. Even though it's more convenient to compare with the timestamps in the filenames of the meta files, other times such as running, stop time are all in seconds. It can be inconsistent.

Reply:
actually other times are in milliseconds too, its only while stored in run records, they are converted to seconds, whereas in the tms its stored as millli-seconds.